### PR TITLE
Implement add_callback_threadsafe in all connection adapters

### LINKED ISSUE #892: Question re. interrupting `BlockingChannel.consume` from another thread
In [Dramatiq][dramatiq] I've got an architecture where consumers live on their own thread and call `BlockingChannel`'s `consume` method with an inactivity timeout of `1s` by default. Whenever a consumer gets a message from Rabbit, it puts it on a shared in-memory work queue where it is grabbed and processed by any of `n` worker threads. Once the worker thread is done processing the message, it gets put on an "acks" queue that belongs to the consumer. When the consumer hits the inactivity timeout it acks any pending messages off its queue and then waits for new messages from RMQ.

Acking from the worker threads is not an option since `BlockingChannel`s are not thread safe.

When you set a QoS value the fact that `consume` can't be interrupted from another thread really starts to hurt throughput, at which point you are forced to either:

1. use a tiny `inactivity_timeout` (<100ms), which "fixes" the throughput issue but is essentially a spinlock so it's inefficient,
2. switch to `SelectConnection` and implement your own interrupt, this is probably what any sane person would do, but the problem is you'd end up implementing a lot of stuff that `BlockingChannel` already implements, or
3. hack your own interrupt by hooking into implementation details of `BlockingChannel`.

The latter is what I ended up doing ([here][c1], [here][c2] and [here][c3]) and my interrupt implementation is working well, but I don't feel too good about depending on `BlockingConnection`'s implementation details so my "issue" is more of a question than anything: is there a better way I could be doing this and, if not, would this be a good addition to `BlockingConnection` (I'd be happy to contribute)? 

[dramatiq]: http://dramatiq.io/
[c1]: https://github.com/Bogdanp/dramatiq/blob/master/dramatiq/brokers/rabbitmq.py#L359-L373
[c2]: https://github.com/Bogdanp/dramatiq/blob/master/dramatiq/brokers/rabbitmq.py#L422-L435
[c3]: https://github.com/Bogdanp/dramatiq/blob/master/dramatiq/brokers/rabbitmq.py#L439-L458

### PR DESCRIPTION
**Pull request by vitaly-krugl**
(*NOTE that the failed coverage test results from pika's encrypted credentials configuration issue*)

This pull request implements a new method `add_callback_threadsafe` on all supported connection adapters in pika. This method enables the application to delegate calls from other threads into the context of the adapter's IOLoop thread (similarly to Tornado's `add_callback`, Twisted's `callFromThread`, and asyncio's `call_soon_threadsafe`).

Lack of this feature has been a pain point for users of `BlockingConnection` and `SelectConnection` for as long as I can remember, and users have had to rely on clumsy and inefficient workarounds.

Here is a sampling of several relevant issues: #963, #927, #892, #930, #752, #418, #753, #636, #564.

For example, this feature facilitates the app to delegate long-lasting processing of incoming messages to a background thread and then use `add_callback_threadsafe` to efficiently delegate ACKing of the message in the context of the connection's thread, thus avoiding lost heartbeat/connection issues from blocking too long and also hacky polling techniques that applications have been using to work around the limitation in the past.

ADDITIONAL INFO FOR CHANGELOG: It is customary among frameworks - such as tornado, twisted, and asyncio - to designate a single threadsafe method that accepts a callback to be executed in the IOLoop's thread, and no others are thread-safe for ease of support/maintenance and for API orthogonality. Following this rationale, I removed the thread-safe designation inside the docstring of the ioloop `stop()` method in `select_connection.IOLoop`.

Fixes #892

### CODE REVIEW

**Comment by vitaly-krugl** [comment]
Anyone have any idea why the Coverage job in travis-ci is always failing in my PR?

cc @michaelklishin, @lukebakken

**Comment by shinji-s** [comment]
The attempt to populate coverage/ directory from s3 storage is failing because credential is not installed/found. Sorry I don't know how to fix though.

**Comment by lukebakken** [comment]
@shinji-s pointed it out -

```
The command "mkdir coverage" exited with 0.
0.49s$ aws s3 cp --recursive s3://com-gavinroy-travis/pika/$TRAVIS_BUILD_NUMBER/ coverage
fatal error: Unable to locate credentials
```

@gmr - would you have time to address this? Thanks!!!

**Comment by shinji-s** [comment]
Regarding the coverage test failure, my guess below may be well off the mark but I'd try.

The credential is given here in .travis.yml
>   \- secure:  {actual-encrypted-string-is-omitted}
>   \- secure:  {actual-encrypted-string-is-omitted}

However, https://docs.travis-ci.com/user/encryption-keys/ has this to say:
> Please note that encrypted environment variables are not available for pull requests from forks.

That is why the credential is not available in the above build?

**Comment by vitaly-krugl** [comment]
@shinji-s, thank you for looking into this. Your assessment makes a ton of sense. If that's the case, then the selected mechanism is inappropriate for the project's needs, since most PRs would naturally come from forks and reviewers need the means to assess impact of the PR on code coverage.

**Comment by lukebakken** [comment]
Reviewing now

File: pika/adapters/select_connection.py [hunk 1]
```diff
@@ -164,6 +163,16 @@ def __init__(self):
         # collection of canceled timeouts
         self._num_cancellations = 0
 
+    def close(self):
```
Comment (lukebakken) [inline pika/adapters/select_connection.py]: Should all of the connection adapters implement `close`? It looks like the Twisted connection `IOLoopReactorAdapter` does not, and the [Tornado docs](http://www.tornadoweb.org/en/stable/ioloop.html#tornado.ioloop.IOLoop.close) state that `stop` must be called before close which isn't enforced in our code. There's also a good chance I'm mis-understanding how these adapters work

Comment (vitaly-krugl) [reply pika/adapters/select_connection.py]: @lukebakken - good points. I will follow up in a couple of days. Thanks!

Comment (vitaly-krugl) [reply pika/adapters/select_connection.py]: @lukebakken, the pull request is ready for resuming code review. Thanks!

Comment (vitaly-krugl) [reply pika/adapters/select_connection.py]: I also discovered the existing issue #970 in the process of implementing new tests.

**Comment by vitaly-krugl** [comment]
@lukebakken - this pull request contains a large number of commits - 39 presently; would it be better if I squashed them into a single commit?

File: pika/adapters/select_connection.py (hunk 1 above)
Comment (lukebakken) [reply pika/adapters/select_connection.py]: Great! I'll get on it today.

**Comment by vitaly-krugl** [comment]
Thanks @lukebakken !

**Comment by lukebakken** [comment]
Great work, this will be very useful for Pika users

**Pull request merged**